File: remote/src/main/java/com/kelsos/mbrc/adapters/AlbumProfileAdapter.java
package com.kelsos.mbrc.adapters;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.PopupMenu;
import android.widget.TextView;
import butterknife.Bind;
import butterknife.ButterKnife;
import com.google.inject.Inject;
import com.kelsos.mbrc.R;
import com.kelsos.mbrc.domain.Track;
import java.util.ArrayList;
import java.util.List;

public class AlbumProfileAdapter extends RecyclerView.Adapter<AlbumProfileAdapter.ViewHolder> {

  private LayoutInflater inflater;
  private List<Track> data;
  private MenuItemSelectedListener listener;

  @Inject public AlbumProfileAdapter(Context context) {
    inflater = LayoutInflater.from(context);
    data = new ArrayList<>();
  }

  public void updateData(List<Track> data) {
    this.data.clear();
    this.data.addAll(data);
    notifyDataSetChanged();
  }

  private void showPopup(View view, Track track) {
    PopupMenu popupMenu = new PopupMenu(view.getContext(), view);
    popupMenu.inflate(R.menu.popup_track);
    popupMenu.setOnMenuItemClickListener(menuItem -> {
      if (listener != null) {
        listener.onMenuItemSelected(menuItem, track);
        return true;
      }
      return false;
    });
    popupMenu.show();
  }

  @Override public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
    final View view = inflater.inflate(R.layout.listitem_single, parent, false);
    ViewHolder holder = new ViewHolder(view);
    holder.overflow.setOnClickListener(v -> {
      int position = holder.getAdapterPosition();
      Track track = data.get(position);
      showPopup(v, track);
    });
    holder.itemView.setOnClickListener(v -> {
      if (listener !=  null) {
        int position = holder.getAdapterPosition();
        Track track = data.get(position);
        listener.onItemClicked(track);
      }
    });
    return holder;
  }

  @Override public void onBindViewHolder(ViewHolder holder, int position) {
    final Track track = data.get(position);
    holder.lineOne.setText(track.getTitle());
  }

  @Override public int getItemCount() {
    return data.size();
  }

  public void setListener(MenuItemSelectedListener listener) {
    this.listener = listener;
  }

  public interface MenuItemSelectedListener {
    void onMenuItemSelected(MenuItem menuItem, Track entry);

    void onItemClicked(Track track);
  }

  public static class ViewHolder extends RecyclerView.ViewHolder {
    @Bind(R.id.line_one) TextView lineOne;
    @Bind(R.id.ui_item_context_indicator) View overflow;

    public ViewHolder(View itemView) {
      super(itemView);
      ButterKnife.bind(this, itemView);
    }
  }
}

File: remote/src/main/kotlin/com/kelsos/mbrc/adapters/AlbumProfileAdapter.kt
package com.kelsos.mbrc.adapters

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.PopupMenu
import android.widget.TextView
import butterknife.BindView
import butterknife.ButterKnife
import com.google.inject.Inject
import com.kelsos.mbrc.R
import com.kelsos.mbrc.domain.Track
import java.util.*

class AlbumProfileAdapter
@Inject constructor(context: Context) : RecyclerView.Adapter<AlbumProfileAdapter.ViewHolder>() {

  private val inflater: LayoutInflater
  private val data: MutableList<Track>
  private var listener: MenuItemSelectedListener? = null

  init {
    inflater = LayoutInflater.from(context)
    data = ArrayList<Track>()
  }

  fun updateData(data: List<Track>) {
    this.data.clear()
    this.data.addAll(data)
    notifyDataSetChanged()
  }

  private fun showPopup(view: View, track: Track) {
    val popupMenu = PopupMenu(view.context, view)
    popupMenu.inflate(R.menu.popup_track)
    popupMenu.setOnMenuItemClickListener listener@{ menuItem ->
      if (listener != null) {
        listener!!.onMenuItemSelected(menuItem, track)
        return@listener true
      }
      false
    }
    popupMenu.show()
  }

  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
    val view = inflater.inflate(R.layout.listitem_single, parent, false)
    val holder = ViewHolder(view)
    holder.overflow.setOnClickListener { v ->
      val position = holder.adapterPosition
      val track = data[position]
      showPopup(v, track)
    }
    holder.itemView.setOnClickListener { v ->
      if (listener != null) {
        val position = holder.adapterPosition
        val track = data[position]
        listener!!.onItemClicked(track)
      }
    }
    return holder
  }

  override fun onBindViewHolder(holder: ViewHolder, position: Int) {
    val track = data[position]
    holder.lineOne.text = track.title
  }

  override fun getItemCount(): Int {
    return data.size
  }

  fun setListener(listener: MenuItemSelectedListener) {
    this.listener = listener
  }

  interface MenuItemSelectedListener {
    fun onMenuItemSelected(menuItem: MenuItem, entry: Track)

    fun onItemClicked(track: Track)
  }

  class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
    @BindView(R.id.line_one) lateinit internal var lineOne: TextView
    @BindView(R.id.ui_item_context_indicator) lateinit internal var overflow: View

    init {
      ButterKnife.bind(this, itemView)
    }
  }
}
